[yixianggg] iP

yixianggg opened the pull request:
### Bob

> “Your mind is for having ideas, not holding them.” – David Allen

Bob frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/yixianggg/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉👍

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `Main` method:
```java
public static void main(String[] args) {
    storage = new Storage("./data/bob.txt");
    taskList = new TaskList();
    ui.showWelcomeMessage(NAME);

    // Load tasks from storage
    try {
        taskList.setTasks(storage.load());
    } catch (ChatBotException e) {
        ui.showError(e.getMessage());
    }

    Scanner scanner = new Scanner(System.in);
    String phrase = scanner.nextLine();

    // Main loop to handle user input
    while (!phrase.equals("bye")) {
        try {
            ui.showLine();
            Parser.handleInput(phrase, taskList, ui);
            storage.save(taskList.getTasks());
        } catch (ChatBotException e) {
            ui.showError(e.getMessage());
        } finally {
            ui.showLine();
            phrase = scanner.nextLine();
        }
    }

    // Show goodbye message when the user exits the program
    ui.showGoodbyeMessage();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

yixianggg committed:
renamed Duke to Bob, implement greet and exit

yixianggg committed:
rename Duke to Bob, implement greet and exit

yixianggg committed:
bob echos commands, closes upon bye command

yixianggg committed:
added lines

yixianggg committed:
added add and list commands

yixianggg committed:
added in Task class, mark and unmark commands for Bob

yixianggg committed:
added new classes Deadline, Event and ToDo, along with their respective commands in Bob

yixianggg committed:
added in expected behaviour and input file

yixianggg committed:
restructured code, split different sections into separate functions. added in exceptions and error handling, updated test cases and expected output

yixianggg committed:
added in delete function, updated test case and expected output

yixianggg committed:
edit syntax and added usage of enums

yixianggg committed:
Fix indentation for switch statement

yixianggg committed:
Add functions to write tasks to file

yixianggg committed:
Add functions to read tasks from file

yixianggg committed:
Merge branch 'branch-Level-7'

yixianggg committed:
Implement understanding of date and times and add function to print out tasks on given date

yixianggg committed:
Merge branch 'branch-Level-8'

yixianggg committed:
Add Ui Class

yixianggg committed:
Add Storage class

yixianggg committed:
Add TaskList class

yixianggg committed:
Add Parser class

yixianggg committed:
Edit classes

yixianggg committed:
Fix adding event code

yixianggg committed:
Merge branch 'branch-A-MoreOOP'

yixianggg committed:
Merge remote-tracking branch 'origin/add-gradle-support'

yixianggg committed:
Edit path in build.gradle

yixianggg committed:
Merge branch 'branch-A-Gradle'

yixianggg committed:
Create bob package

yixianggg committed:
Merge branch 'branch-A-Packages'

yixianggg committed:
Bug fixes

yixianggg committed:
Add TaskList tests

yixianggg committed:
Add Parser tests

yixianggg committed:
Merge branch 'branch-A-JUnit'

yixianggg committed:
Edit build.gradle

yixianggg committed:
Merge branch 'branch-A-Jar'

yixianggg committed:
Add javadoc for Bob class

yixianggg committed:
Add javadoc for ChatBotException class

yixianggg committed:
Add javadoc for Deadline class

yixianggg committed:
Add javadoc for Event class

yixianggg committed:
Add javadoc for Parser class

yixianggg committed:
Add javadoc for storage class

yixianggg committed:
Add javadoc for Task class

yixianggg committed:
Add javadoc for TaskList class

yixianggg committed:
Add javadoc for TaskType enum

yixianggg committed:
Add javadoc for ToDo class

yixianggg committed:
Add javadoc for Ui class

yixianggg committed:
Add javadoc for ParserTest class

yixianggg committed:
Add javadoc for TaskListTest class

yixianggg committed:
Edit code to follow coding standards

yixianggg committed:
Add find function

yixianggg committed:
Add javadoc for find function

yixianggg committed:
Merge branch 'branch-A-JavaDoc'

yixianggg committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	data/bob.txt
#	src/main/java/bob/Event.java

yixianggg committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	data/bob.txt
#	src/main/java/bob/Parser.java
#	src/main/java/bob/TaskList.java
#	src/main/java/bob/Ui.java

yixianggg committed:
Add checkstyle plugin and fix style errors

yixianggg committed:
Merge branch 'branch-A-CheckStyle'

yixianggg committed:
Follow javafx tutorials 1 and 2

yixianggg committed:
Follow javafx tutorials 3 and 4

yixianggg committed:
Merge branch 'branch-Level-10'

yixianggg committed:
Edit showTasksFound method in Ui class to use Java varargs

yixianggg committed:
Merge branch 'branch-A-Varargs'

LingZiyann commented on src/main/java/bob/Deadline.java:
```diff
@@ -0,0 +1,85 @@
+        return this.by;
+    }
+
+    /**
+     * Returns a string representation of the deadline task, including its description and formatted deadline date/time.
+     * This method overrides the {@link Task#toString()} method to provide a specific format for Deadline tasks.
+     *
+     * @return The string representation of the deadline task.
+     * @inheritDoc
+     */
+    @Override
```
Writing javadocs for overridden comments are actually optional. Just pointing out in case you didn't know

LingZiyann commented on src/main/java/bob/TaskList.java:
```diff
@@ -0,0 +1,284 @@
+     *
+     * @return The number of tasks in the TaskList.
+     */
+    public int size() {
```
method name could be more descriptive

LingZiyann commented on src/main/resources/images/DaBob.png:
Nice image

LingZiyann commented on src/main/java/bob/MainWindow.java:
```diff
@@ -0,0 +1,109 @@
+     * If the user input is "bye", displays the goodbye message and closes the application after a short delay.
+     */
+    @FXML
+    private void handleUserInput() {
```
I like how u made use of comments to describe the inner logic, but some comments could perhaps be omitted as it is trivial

LingZiyann commented on src/main/java/bob/TaskType.java:
```diff
@@ -0,0 +1,21 @@
+/**
+ * Represents the different types of tasks that can be managed in the task list.
+ */
+public enum TaskType {
```
nice comments to explain the Enum types here

LingZiyann commented on src/main/java/bob/Ui.java:
```diff
@@ -0,0 +1,112 @@
+     * @param botName The name of the chatbot.
+     * @return The formatted welcome message.
+     */
+    public String showWelcomeMessage(String botName) {
```
i like the method namings here

LingZiyann reviewed (commented):
I think pretty much everything adheres to the java coding standards given. Good job

maahir-garg reviewed (commented):
Overall looks great, maybe just think about how to further divide your classes into packages like commands, tasks etc

yixianggg committed:
The TaskList class does not currently validate assumptions about task
list size after adding or removing tasks, which could lead to
undetected bugs during development.

Without assertions, there is no immediate check on the expected
behaviour of task list operations. This could allow incorrect task
management logic to go unnoticed, leading to potential runtime errors
or logical inconsistencies.

Add assertions in the addToDo and deleteTask methods to verify that the
task list size changes as expected after adding or removing tasks.
Also, include an assertion to check that the phrase length for a to-do
task is valid before proceeding with task creation.

Assertions are used to document assumptions and catch errors during the
development phase. By ensuring that these conditions are met, we can
prevent logical errors from propagating and ensure the integrity of the
task list operations.

yixianggg committed:
Currently, the task-saving process uses a traditional for-each loop to
write tasks to a file, leading to more verbose code.

This needs to change because the code can be made more concise and
expressive by leveraging Java's Streams API, which allows for a more
functional and readable approach to transforming collections.

Refactor the save method to use streams to map tasks to their string
representations and collect them into a single output for the file
writer.

This is done to enhance readability, eliminate manual iteration, and
provide a more declarative way to handle collections in Java. The
Streams API is more efficient and idiomatic for tasks like these.

yixianggg committed:
Remove unnecessary comments

yixianggg committed:
The file existence checks in the load and save methods are written
directly inside if conditions, making the code less readable and
slightly harder to maintain.

Refactoring the conditions into separate boolean variables improves
readability and makes the logic easier to understand, debug, and
maintain in the future. This structure also reduces repetition within
the methods and provides clarity in understanding which conditions are
being checked.

Extract boolean expressions for checking file and directory existence
into named variables (directoryFileExists, directoryExists, fileExists,
and isRegularFile). This clarifies the logic and organizes the code for
better readability and maintainability.

By separating the conditions into clearly named variables, we ensure
that the logic is easy to follow and modify. It also promotes clean
code practices by reducing complex conditional expressions.

yixianggg committed:
Merge pull request #1 from yixianggg/branch-A-Assertions

Add assertions to TaskList class

yixianggg committed:
Merge branch 'master' into branch-A-Streams

yixianggg committed:
Merge branch 'master' into branch-A-CodeQuality

yixianggg committed:
Merge pull request #2 from yixianggg/branch-A-Streams

Use streams in Save method

yixianggg committed:
Merge branch 'master' into branch-A-CodeQuality

yixianggg committed:
Merge pull request #3 from yixianggg/branch-A-CodeQuality

Improve code readability in Storage class

yixianggg committed:
Add .yml file

yixianggg committed:
Add empty line after .yml file

yixianggg committed:
Remove unnecessary comments

yixianggg committed:
Edit .yml file to change checkstyle files

yixianggg committed:
Remove trailing whitespace in Line 276

yixianggg committed:
Merge branch 'branch-A-CI'

yixianggg committed:
Add in new task type - Fixed, which accommodates tasks with no deadline
but have a fixed duration to complete

yixianggg committed:
Merge branch 'branch-B-FixedDurationTasks'

yixianggg committed:
Rename boolean variable found to isFound in searchKeyword method

searchKeyword method uses found as a boolean variable - uses unclear
naming for boolean values

Boolean variables should have names that indicate true/false values
more clearly.

Rename found to isFound for better clarity and adherence to boolean
naming conventions.

This change improves readability and aligns with standard coding
practices for boolean variable names.

yixianggg committed:
Remove unused constructor from TaskList

The constructor in the TaskList class accepts an existing list of tasks.

This constructor is currently not being used in the codebase.

Remove the constructor to eliminate unused code and simplify the
TaskList class.

By removing this unused constructor, we ensure the codebase remains
clean, maintainable, and free of unnecessary functionality.

yixianggg committed:
Remove showLine method from Ui class

The showLine method in the Ui class returns a horizontal line for
formatting output.

This method is redundant and no longer contributes to the UI's
functionality.

Remove the showLine method to clean up the UI class and avoid
unnecessary methods.

By removing this method, we streamline the UI class, ensuring that only
relevant and essential functionality is retained.

yixianggg committed:
Remove getWelcomeMessage method from Bob class

The getWelcomeMessage method in the Bob class provides a welcome
message when the chatbot starts.

This method is redundant because a similar method exists in the Ui
class, which already handles welcome messages.

Remove the getWelcomeMessage method from the Bob class to avoid
duplication and rely on the method in the Ui class for consistency.

By removing this method, we maintain a single source of truth for
displaying welcome messages and reduce code redundancy.

yixianggg committed:
Remove getGoodbyeMessage method from Bob class

The getGoodbyeMessage method in the Bob class provides a goodbye
message when the user exits the chatbot.

This method is redundant because a similar method exists in the Ui
class, which already handles goodbye messages.

Remove the getGoodbyeMessage method from the Bob class to avoid
duplication and use the method in the Ui class for consistency.

This reduces code redundancy and ensures that the farewell messages are
managed centrally in the Ui class.

yixianggg committed:
Add user guide to README

Currently, there is no user guide in the project’s documentation, which
leaves users unclear on how to interact with the Bob chatbot.

This needs to change to ensure that users have clear instructions on
how to use the chatbot, including its commands and features.

Add a detailed user guide under the README, outlining key features like
adding tasks, marking tasks, finding tasks, and exiting Bob. Provide
instructions for different task types, task management, and usage
examples.

This is done to help users easily understand and navigate the
functionalities of Bob, improving overall user experience.

yixianggg committed:
Set application window title to "Bob Chatbot"

The application window currently displays no title, making it unclear
what the window represents.

This needs to change so that users can easily identify the application
they are interacting with.

Set the title of the primary stage to "Bob Chatbot."

It ensures clarity and enhances the user's experience by showing the
chatbot's name in the title bar.

yixianggg committed:
Fix bug by including isDone in Fixed constructor's super call

The Fixed constructor does not reference the isDone
parameter when calling the superclass constructor, causing fixed
tasks to always be read as unmarked.

This needs to change to correctly reflect the completion status of
fixed tasks.

Include the isDone parameter in the call to the superclass
constructor.

It ensures that the fixed tasks correctly maintain and display their
completion status.

yixianggg committed:
Add screenshot of Bob Chatbot UI to user guide

The user guide does not currently include a visual representation of
the Bob Chatbot UI.

Including a screenshot helps users visualize the application's
interface and understand its features better.

Add a representative screenshot of the Bob Chatbot UI to the user guide.

It enhances the documentation by providing a clear view of the user
interface, making it more informative and user-friendly.

yixianggg committed:
Merge branch 'branch-A-UserGuide'

yixianggg committed:
Update user guide to replace * with ! emoji for emphasis

The user guide uses * for emphasis in Feature section, which can be
less noticeable.

Replace * with the ! emoji to improve the visibility of important
notes.

Replace * with the ! emoji in the user guide to highlight crucial
information regarding date formatting.

It improves readability and draws more attention to important details,
enhancing user experience.

yixianggg committed:
Merge branch 'branch-A-UserGuide'

yixianggg committed:
Update user guide to use * for italicization instead of _

The user guide uses _ for italicizing text, which does not work as
intended in Markdown.

_ does not italicize in Markdown, hence not achieving the intended
formatting.

Change _ to * to ensure proper italicization of text in the user guide.

Adjust italicization in the user guide to use * instead of _ to
correctly format text and improve readability.

yixianggg committed:
Merge branch 'branch-A-UserGuide'

yixianggg committed:
Update user guide to not use italicization

The user guide uses * for italicizing text, which does not work as
intended in Markdown.

* does not italicize in Markdown, hence not achieving the intended
formatting.

Remove the instance of italicization in the user guide.

Remove usage of italicization to remove instances of * that could
mislead readers and affect readability.

yixianggg committed:
Merge branch 'branch-A-UserGuide'

yixianggg committed:
Fix date formatting display in user guide

The user guide currently uses angle brackets <date> and
<yyyy-MM-dd HHmm> for date formatting instructions.

Angle brackets do not render properly in Markdown, causing the
instructions to appear incorrectly.

Replace angle brackets with quoting codes (`<date>` and
`<yyyy-MM-dd HHmm>`) to ensure proper display.

This change improves readability and ensures that formatting
instructions are clear and accurate, enhancing user understanding and
adherence to the date formatting guidelines.

yixianggg committed:
Merge branch 'branch-A-UserGuide'

yixianggg committed:
Add exception handling and assertions to task addition methods

The addToDo method did not handle invalid input properly and lacked
internal consistency checks.

To improve error handling and ensure the integrity of task addition
operations.

Added validation to ensure the to-do description is not empty and
throws a ChatBotException for invalid inputs. Incorporated assertions
to check that the task list size increases correctly after adding a new
task. Similar changes were made to the addDeadline, addEvent, and
addFixed methods to include assertions and improve robustness.

These updates ensure that invalid inputs are properly managed and
internal consistency is maintained across task addition operations.
